leakpro/attacks/gia_attacks/invertinggradients.py

```diff
@@ -0,0 +1,167 @@
+        optimizer = torch.optim.Adam([self.reconstruction], lr=self.attack_lr)
+        scheduler = torch.optim.lr_scheduler.MultiStepLR(optimizer,
+                                                            milestones=[24000 // 2.667, 24000 // 1.6,
+
+                                                                        24000 // 1.142], gamma=0.1)
+
+        for i in range(500):
+            closure = self.gradient_closure(optimizer)
+
+            loss = optimizer.step(closure)
+            scheduler.step()
+            with torch.no_grad():
+                self.reconstruction.data = torch.max(
+                    torch.min(self.reconstruction, (1 - self.data_mean) / self.data_std), -self.data_mean / self.data_std
+                    )
+            if i % 20 == 0:
+                self.logger.info(f"{i}: {loss}")
+            # add PSNR calculation and pick best image..
+        # Collect client data to one tensor
+        all_data = []
+
+        for i in range(len(self.client_loader.dataset)):
+            x, _ = self.client_loader.dataset[i]
+            all_data.append(x)
+
+        # Convert lists to tensors
+        original_data_tensor = torch.stack(all_data)
+        return GIAResults(original_data_tensor, self.reconstruction, 0, self.data_mean, self.data_std)
+
+    def total_variation(self: Self, x: torch.Tensor) -> torch.Tensor:
+        """Anisotropic TV."""
+        dx = torch.mean(torch.abs(x[:, :, :, :-1] - x[:, :, :, 1:]))
+        dy = torch.mean(torch.abs(x[:, :, :-1, :] - x[:, :, 1:, :]))
+        return dx + dy
+
+
+    def gradient_closure(self: Self, optimizer: torch.optim.Optimizer) -> Callable:
+        """Returns a closure function that performs a gradient descent step.
+
+        The closure function computes the gradients, calculates the reconstruction loss,
+        adds a total variation regularization term, and then performs backpropagation.
+        """
+        def closure() -> torch.Tensor:
+            """Computes the reconstruction loss and performs backpropagation.
+
+            This function zeroes out the gradients of the optimizer and the model,
+            computes the gradient and reconstruction loss, logs the reconstruction loss,
+            optionally adds a total variation term, performs backpropagation, and optionally
+            modifies the gradient of the input image.
+
+            Returns
+            -------
+                torch.Tensor: The reconstruction loss.
+
+            """
+            optimizer.zero_grad()
+            self.handler.target_model.zero_grad()
+
+            gradient = self.handler.train(self.reconstruction_loader, self.handler.get_optimizer())
+            rec_loss = self.reconstruction_costs([gradient], self.client_gradient)
+
+            self.logger.info(f"rec loss {rec_loss}")
+            rec_loss += self.t_v_scale * self.total_variation(self.reconstruction)
+            rec_loss.backward()
+            self.reconstruction.grad.sign_()
```

Comment:
what does this do?

Reply:
added a comment:
"# Add the TV loss term to penalize large variations between pixels, encouraging smoother images."